### PR TITLE
Add ContravariantCartesian typeclass and for Tuple2 add Monad and FlatMap

### PR DESCRIPTION
**Pull request by johnynek**
These new instances should be isomorphic to `WriterT[Id, A, B]` instances, but there may be cases where you already have a tuple2 and want to use a monad.

This approach can be generalized to `TupleN`, but I am not sure it yet justifies code-gen or macros for this purpose.

Interestingly, one could make a `Monad` for a general case-class that is generic in at least the last parameter, but that may be going too far.

I also added `Cartesian[Eq]` to make the tests easier (otherwise I needed to plumb some new implicits everywhere).

### CODE REVIEW

**Comment by striation** [comment]
LGTM 👍

File: core/src/main/scala/cats/instances/tuple.scala [hunk 1]
```diff
@@ -1,9 +1,12 @@
 package cats
 package instances
 
+import data.Xor
```
Comment (adelbertc) [inline core/src/main/scala/cats/instances/tuple.scala]: Either! :D

**Comment by adelbertc** [comment]
https://github.com/typelevel/cats/pull/1289 is merged now - looks like this change doesn't have a merge conflict but I suspect if merged it will break the build :)

**Comment by johnynek** [comment]
I'll update, get the coverage positive and try again.

**Comment by johnynek** [comment]
I think this is green now, except for the PR failing because of the `Xor => Either`. Since that is being reverted, this should be green after #1305 goes through.

**Comment by adelbertc** [comment]
:+1:

**Comment by kailuowang** [comment]
👍

**Pull request merged**